plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -296,130 +271,53 @@ function EpubDownloadBackend:createEpub(epub_path, html, url, include_images, me
-<package xmlns="http://www.idpf.org/2007/opf"
-        xmlns:dc="http://purl.org/dc/elements/1.1/"
-        unique-identifier="bookid" version="2.0">
-  <metadata>
-    <dc:title>%s</dc:title>
-    <dc:publisher>KOReader %s</dc:publisher>
-    %s
-  </metadata>
-  <manifest>
-    <item id="ncx" href="toc.ncx" media-type="application/x-dtbncx+xml"/>
-    <item id="content" href="content.html" media-type="application/xhtml+xml"/>
-    <item id="css" href="stylesheet.css" media-type="text/css"/>
-]], page_htmltitle, Version:getCurrentRevision(), meta_cover))
-    -- images files
-    if include_images then
-        for inum, img in ipairs(images) do
-            table.insert(content_opf_parts, string.format([[    <item id="%s" href="%s" media-type="%s"/>%s]], img.imgid, img.imgpath, img.mimetype, "\n"))
-        end
-    end
-    -- tail
-    table.insert(content_opf_parts, [[
-  </manifest>
-  <spine toc="ncx">
-    <itemref idref="content"/>
-  </spine>
-</package>
-]])
-    epub:add("OEBPS/content.opf", table.concat(content_opf_parts))
-    logger.dbg("Added OEBPS/content.opf")
-
-    -- ----------------------------------------------------------------
-    -- OEBPS/stylesheet.css
-    --- @todo We told it we'd include a stylesheet.css, so it's probably best
-    -- that we do. In theory, we could try to fetch any *.css files linked in
-    -- the main html.
-    epub:add("OEBPS/stylesheet.css", [[
-/* Empty */
-]])
-    logger.dbg("Added OEBPS/stylesheet.css")
-
-    -- ----------------------------------------------------------------
-    -- OEBPS/toc.ncx : table of content
+function EpubBuilder:addToc(chapters)
     local toc_ncx_parts = {}
     local depth = 0
-    local cur_level = 0
-    local np_end = [[</navPoint>]]
-    local num = 1
-    -- Add our own first section for first page, with page name as title
-    table.insert(toc_ncx_parts, string.format([[<navPoint id="navpoint-%s" playOrder="%s"><navLabel><text>%s</text></navLabel><content src="content.html"/>]], num, num, page_htmltitle))
-    table.insert(toc_ncx_parts, np_end)
-    --- @todo Not essential for most articles, but longer articles might benefit
-    -- from parsing <h*> tags and constructing a proper TOC
-    while cur_level > 0 do
-        table.insert(toc_ncx_parts, np_end)
-        cur_level = cur_level - 1
+    local num = 0
+
+    for index, chapter in ipairs(chapters) do
+        -- Add nav part for each chapter.
+        table.insert(
+            toc_ncx_parts,
+            string.format([[<navPoint id="navpoint-%s" playOrder="%s"><navLabel><text>%s</text></navLabel><content src="%s.html"/></navPoint>]],
+                num,
+                num,
+                chapter.title,
+                chapter.md5
+            )
+        )
+        num = num + 1
     end
-    -- Prepend NCX head
-    table.insert(toc_ncx_parts, 1, string.format([[
+    -- Prepend NCX head.
+    table.insert(
+        toc_ncx_parts,
+        1,
+        string.format([[
```

Comment:
Maybe it's just me, but I find this a bit hard to read. (A bit of a broader comment, but this one in particular stands out.)

Reply:
Yeah, totally agree. Those blocks with the XML are all very ugly... I'm open to suggestions :)